File: problems/problems_124/solution.ts
import {TreeNode,JSONArrayToTreeNode} from "../../typescript/models/treenode";

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxPathSum(root: TreeNode | null): number {
	let ans: number = Number.MIN_SAFE_INTEGER;
	const dfs = (node: TreeNode | null): number => {
		if (node == null) {
			return 0;
		}
		const left: number = dfs(node.left);
		const right: number = dfs(node.right);
		ans = Math.max(ans, node.val + left + right);
		return Math.max(0, node.val + Math.max(left, right));
	}
	dfs(root);
	return ans;
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const root: TreeNode | null = JSONArrayToTreeNode(JSON.parse(inputValues[0]));
	return maxPathSum(root);
}

File: problems/problems_124/solution.rs
use serde_json::{json, Value};
use library::lib::tree_node::{TreeNode, array_to_tree};
pub struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn max_path_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
		let mut ans: i32 = i32::MIN;
		fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, ans: &mut i32) -> i32 {
			if let Some(node) = node {
				let left = dfs(&node.borrow().left, ans);
				let right = dfs(&node.borrow().right, ans);
				*ans = (*ans).max(node.borrow().val + left + right);
				0.max(node.borrow().val + left.max(right))
			} else {
				0
			}
		}
		dfs(&root, &mut ans);
		ans
    }
}

#[cfg(feature = "solution_124")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let input_vec0: Vec<Option<i32>> = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	let root: Option<Rc<RefCell<TreeNode>>> = array_to_tree(&input_vec0);
	json!(Solution::max_path_sum(root))
}
